r/R/dotplot.R

```diff
@@ -1,35 +1,84 @@
+    return(list())
+  }
 
-  for (i in seq_along(cell_sets)) {
-    set <- cell_sets[[i]]
-    filtered_cells <- intersect(set$cellIds, cells_id)
-    data$custom[cells_id %in% filtered_cells] <- i
+  #Collect ids to subset object
+  if (apply_filter) {
+    subsetIds <- filter_by$cellIds
+  } else {
+    subsetIds <- list()
+    for (child in filter_by$children) {
+      subsetIds <- append(subsetIds, child$cellIds)
+    }
   }
 
-  if (type == "custom") {
-    annot <- data@misc$gene_annotations
-    annot_subset <- subset(annot, toupper(name) %in% toupper(req_genes))
-    annot_subset <- annot_subset[, c("input", "name")]
+  #Subset cell Ids
+  if (length(subsetIds)) {
+    data <- subset_ids(data,subsetIds)
+    cells_id <- data$cells_id
   } else {
-    # type == marker
-    # Need the marker features
+    message("The ids to subset the object are empty. Returning empty results.")
+    return(list())
   }
 
-  dotplot_data <- Seurat::DotPlot(data, features = annot_subset$input, group.by = "custom")$data
-  #features.plot has the ensemble ids
-  dotplot_data$name <- annot_subset[dotplot_data$features.plot, "name"]
+  #Construct the dotplot.groups slot
+  for (i in seq_along(group_by_cell_sets)) {
+    cell_set <- group_by_cell_sets[[i]]
+    filtered_cells <- intersect(cell_set$cellIds, cells_id)
+    #This covers a border case where two cell_sets have the same name (but different ID). Can happen in scratchpad
+    if (cell_set$name %in% data$dotplot.groups) {
+      data$dotplot.groups[cells_id %in% filtered_cells] <- paste0(cell_set$name, i)
+    } else {
+      data$dotplot.groups[cells_id %in% filtered_cells] <- cell_set$name
+    }
+  }
+  # If NA values are left in the group, dotplot function will fail.
+  data <- subset(data, subset = dotplot.groups != "NA")
 
-  dotplot_data <- dotplot_data[, c("avg.exp", "pct.exp", "name", "id")]
-  colnames(dotplot_data) <- c("avgExpression", "cellsPercentage", "geneName", "cellSet")
+  #Get marker genes or requested gene names.
+  if (useMarkerGenes) {
+    num_features <- req$body$numberOfMarkers
+    all_markers <- getTopMarkerGenes(num_features, data, group_by_cell_sets)
+    features <- as.data.frame(getMarkerNames(data, all_markers))
+    rownames(features) <- features$input
+  } else {
+    req_genes <- req$body$customGenesList
+    annot <- data@misc$gene_annotations
+    annot_subset <- subset(annot, toupper(name) %in% toupper(req_genes))
+    features <- annot_subset[, c("input", "name")]
+  }
 
-  res <- purrr::transpose(dotplot_data)
+  dotplot_data <- Seurat::DotPlot(data, features = features$input, group.by = "dotplot.groups")$data
+  # features.plot has the ensemble ids
+  dotplot_data$name <- features[dotplot_data$features.plot, "name"]
+  dotplot_data <- dotplot_data[stringr::str_order(dotplot_data$id, numeric = TRUE), ]
+  dotplot_data <- dotplot_data %>% transmute(cellSets = as.character(id), geneName = as.character(name), avgExpression = avg.exp, cellsPercentage = pct.exp)
```

Comment:
To me it looks like Seurat is using `avg.exp.scaled` column for the dot plot

Reply:
I had originally changed to avg.exp.scaled but switched back